Add "@asset" to decorate a function as a DAG and an asset

uranusjr opened the pull request:
## Why
As part of [AIP-75](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=311627076#AIP75NewAssetCentricSyntax-ReferenceanAsset)

## What
Add an `asset` decorator which can be used to decorate a function and create a DAG and an asset.

```python
@asset(uri="s3://bucket/object", schedule=None)
def asset2_producer(self, context, asset1_producer):
    print(self)
    print(context["inlet_events"][asset1_producer])
```

In the example above, it create a `DAG(dag_id="asset2_producer", schedule=None, ...)` and an `Asset(name="asset2_producer", uri="s3://bucket/object")`. Inside the function, `self` can be used to access the asset with name `asset2_producer`, `context` is just like `context` in a normal task and `asset1_producer` is used to access an asset with name=asset1_producer


Closes: https://github.com/apache/airflow/issues/42314

uranusjr committed:
Implement asset definition creating a DAG

uranusjr committed:
Basic inlet dependency

uranusjr committed:
Make AssetDefinition subclass Asset

This seems to be the best way for 'schedule' dependencies to work. Still
not entirely sure; we'll revisit this.

Lee-W committed:
style: fix mypy error

Lee-W committed:
feat(asset): allow uri to be None

Lee-W committed:
fix: temporarily serialize AssetDefintion into a string

Lee-W committed:
feat(decorators/assets): rewrite how asset definition is serialized

Lee-W committed:
test(decorators/assets): add test cases to check whether asset decorator generate the right asset definition

Lee-W committed:
test(decorators/assets): add test cases to AssetDefinition

Lee-W committed:
test(decorators/asset): add test cases to Test_AssetMainOperator

Lee-W committed:
test(decorators/assets): remove unused fixtures

Lee-W committed:
docs(example_dag): add example dag for asset_decorator

Lee-W commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,129 @@
+            elif key == "context":
+                value = context
+            else:
+                # TODO: This does not check if the upstream asset actually
```
I think we should simply do nothing. It seems to be the current behavior of our inlets if they're not referenced

uranusjr replied:
Makes sense.

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,129 @@
+                    Asset(name=inlet_aset_name)
+                    for inlet_aset_name in parameters
+                    if inlet_aset_name not in ("self", "context")
+                ],
```
I don’t think this is good enough since `Asset(name="a")` and `Asset(name="a", uri="somewhere")` are considered different assets. Here we only have the name, so maybe one way to do this would be to add a stub (AssetRef?) here, and do a second pass in the DAG processor to get the actual assets? Or maybe we should just keep the AssetRef here, and add code to task execution at runtime to resolve it into an actual asset, similar to how we handle AssetAlias.

Lee-W replied:
~~What are the `actual assets` you mentioned? What should we reference assets if we're using the name `a`?~~

~~1. `Asset(name="a")` only~~
~~2. All the assets with `name=a` (e.g, `Asset(uri="a")`, `Asset(uri="s3://...", name="a")`, `Asset(uri="gcs://...", name="a")`)~~

Lee-W replied:
Just traced the code a bit and figure out we've not yet handled it in inlet side https://github.com/apache/airflow/blob/06088a3abcbb46533e74de360746db766d50cf66/airflow/utils/context.py#L242

Lee-W replied:
```python
@asset(uri="snowflake://.../bus_trips")
def aggregated_bus_trips(raw_bus_trips, context):
    # Write aggregated bus trips to asset...
    print(context["inlet_events"][raw_bus_trips])
```

@uranusjr 
Would also like to confirm by defining something above, we want to get the asset_event of the active asset with the name `raw_bus_trips`. Or is there anything I missed?

Lee-W replied:
namely we'll probably need to change the SQL here https://github.com/apache/airflow/blob/06088a3abcbb46533e74de360746db766d50cf66/airflow/utils/context.py#L291

Lee-W replied:
also you mentioned the DAG processing thing, is that still needed if I go with the above approach (I guess not?)

uranusjr replied:
Yeah this should work for the user. Although I feel we should resolve `raw_bus_trips` inside the function into an Asset object automatically when we execute it, instead of having it as a ref and do special deferecencing inside the accessor. This is because the user may want to use the value for other purposes, and it would likely be more difficult to make things make sense if `raw_bus_trips` is of a different class than Asset, which is the primary interface to the user.

Lee-W replied:
If that's the case, we probably should just fetch the active asset here https://github.com/apache/airflow/pull/41325/files#diff-1331de0aa19a45db3dbaaa3aab5cd28a00e9090fc33dfa20d4a7fe75887cb1efR93 ?

If we resolve somewhere around execution https://github.com/apache/airflow/pull/41325/files#diff-1331de0aa19a45db3dbaaa3aab5cd28a00e9090fc33dfa20d4a7fe75887cb1efR66, the `inlets` would still be `AssetRef`. Then we'll need to do it here and also do the deferecencing thing.

uranusjr replied:
The values in `inlets` being AssetRef is probably OK; introspecting a task is not a common thing to do (we need to support things like OpenLineage, but that’s not too difficult). We can probably resolve as late as `determine_kwargs`.

Lee-W replied:
I'm now resolving it in `determine_kwargs` and `InletEventsAccessors.__getitem__`. The `InletEventsAccessors.__getitem__` can probably be cached but it's not done at this moment.

Lee-W replied:
I'm thinking of checking the outlet and inlet thing again in another PR. This might be handled there.

Lee-W committed:
feat(decorators/assets): allow passing self and context into asset

Lee-W committed:
feat(decorators/assets): return actual asset in asset decorator

Lee-W committed:
refactor(decorators/assets): extract active assets fetching logic as _fetch_active_assets_by_name

Lee-W committed:
feat(decorators/assets): allow fethcing inlet events through AssetRef

Lee-W committed:
feat(decorators/assets): reorder import paths

Lee-W committed:
docs: update asset decorator example dag

Lee-W committed:
test: fix tests

Lee-W committed:
test(decorators/assets): extend test_determine_kwargs to cover active asset

uranusjr commented on airflow/assets/__init__.py:
```diff
@@ -24,6 +24,7 @@
 from typing import TYPE_CHECKING, Any, Callable, ClassVar, Iterable, Iterator, cast, overload
 
 import attr
+import attrs
```
Let’s unify these, they are the same lib and we should only import from one of them at a time. (Does not really matter which; `attrs` is probably better if we only use the type annotation style.)

https://www.attrs.org/en/stable/names.html

Lee-W replied:
Yep, I was confused a bit and decided to check it later and forgot... thanks for reminding me. will update it.

uranusjr commented on airflow/assets/__init__.py:
```diff
@@ -378,12 +386,18 @@ class Dataset(Asset):
     asset_type: ClassVar[str] = "dataset"
 
+    def __init__(self, *args, group="dataset", **kwargs) -> None:
+        super().__init__(*args, group=group, **kwargs)
```
I think we should make this smarter in the base class so subclasses’ `group` default to `asset_type` automatically. Requiring all subclasses to implement `__init__` like this is problematic.

Lee-W replied:
yep, it's mostly likely due to rebasing error 🤦‍♂️ will fix it

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,159 @@
+
+import inspect
+import types
+from collections.abc import Callable
```
Maybe import from `typing` instead of consistency? Or alternatively import Iterator, Mapping, and Sequence from `collections.abc` instead.

Lee-W replied:
sounds good. will update it

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,159 @@
+        value: Any
+        for key in inspect.signature(self.python_callable).parameters:
+            if key == "self":
+                key = "_self"
```
Why do we need to change the key here?

Lee-W replied:
Because there'll be multiple self here https://github.com/apache/airflow/blob/ff6038b8d154ccdde65a4ebd652ad41d88e8c33e/providers/src/airflow/providers/standard/operators/python.py#L210

uranusjr replied:
Wait what… do you have an example?

Lee-W replied:
```python
from __future__ import annotations

import pendulum

from airflow.assets import Asset
from airflow.decorators import dag, task
from airflow.decorators.assets import asset

@asset(uri="s3://bucket/asset1_producer", schedule=None)
def asset1_producer():
    pass

@asset(uri="s3://bucket/object", schedule=None)
def asset2_producer(self, context, asset1_producer):
    print(self)
    print(context["inlet_events"][asset1_producer][-1])

@dag(
    schedule=Asset(uri="s3://bucket/asset1_producer", name="asset1_producer")
    | Asset(uri="s3://bucket/object", name="asset2_producer"),
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    tags=["consumes", "asset-scheduled"],
)
def consumes_asset_decorator():
    @task(outlets=[Asset(uri="s3://consuming_1_task/asset_other.txt", name="endpoint")])
    def process_nothing():
        pass

    process_nothing()

consumes_asset_decorator()
```

When I remove this `_self` line and comment out the `_handle_self_argument` function, I encounter the following error

```
Traceback (most recent call last):
  File "/opt/airflow/airflow/models/taskinstance.py", line 759, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
  File "/opt/airflow/airflow/models/taskinstance.py", line 725, in _execute_callable
    return ExecutionCallableRunner(
  File "/opt/airflow/airflow/utils/operator_helpers.py", line 258, in run
    return self.func(*args, **kwargs)
  File "/opt/airflow/airflow/models/baseoperator.py", line 375, in wrapper
    return func(self, *args, **kwargs)
  File "/opt/airflow/providers/src/airflow/providers/standard/operators/python.py", line 192, in execute
    return_value = self.execute_callable()
  File "/opt/airflow/providers/src/airflow/providers/standard/operators/python.py", line 210, in execute_callable
    return runner.run(*self.op_args, **self.op_kwargs)
TypeError: run() got multiple values for argument 'self'
```

uranusjr replied:
Ah… OK that makes sense. I’m not really keen on this rename though (especially user can name an asset `_self` and break things). We should be able to work around this with positional arguments. Let me try…

uranusjr replied:
Actually… I can just make the runner work better.

Lee-W replied:
should we include that in this PR? or make it a separate one?

uranusjr replied:
I’m doing that separately.

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,159 @@
+            asset_names.append(self._definition_name)
+
+        if asset_names:
+            self._active_assets = _fetch_active_assets_by_name(asset_names)
```
Do we need to keep the dict as a member? I feel we can simply pass the dict into `_iter_kwargs` instead.

Lee-W replied:
nope, I though that's inherited, but just checked and confirmed it's not

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,159 @@
+        if (name := f.__name__) != f.__qualname__:
+            raise ValueError("nested function not supported")
+        if name == "self" or name == "context":
+            raise ValueError(f"prohibited name for asset: {name}")
```
Doing `Asset(name="self")` (and `context`) directly should also be banned. We should move these checks to `Asset` instead.

Lee-W replied:
Sounds good. will do.

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,159 @@
+        kw_only=True,
+        default="",
+        validator=[attrs.validators.max_len(1500), _validate_identifier],
+    )
```
Probably don’t need these anymore since the validation is already done in `Asset`.

uranusjr replied:
Hmm actually failing fast here is not a bad idea. We should think of a way to keep the validation in sync here and in Asset though.

Lee-W replied:
The only thing I can think of is probably adding a unit test? They already using the same `_validate_identifier` function and `@asset` doesn't have an asset_type by default to fallback to

Lee-W replied:
After a second thought, it's probably not as useful for failing fast here. The only thing we do before the check is check the nested function.

uranusjr commented on airflow/models/asset.py:
```diff
@@ -37,8 +39,30 @@
+            .join(AssetActive, AssetActive.name == AssetModel.name)
+            .where(AssetActive.name.in_(name for name in names))
+        )
+    }
```
```suggestion
    return {
        model.name: model.to_public()
        for model in session.execute(
            select(AssetModel).where(AssetModel.name.in_(names), AssetModel.active.has())
        )
    }
```

Lee-W replied:
Oh, I thought we didn't want to expose anything model-related and didn't have this method, but it turns out we did have to_public. Thanks!

Lee-W committed:
fix: address easy to fix comments

Lee-W committed:
fix: fix asset serialization

Lee-W committed:
refactor(decorators/assets): postpone the attribute check to AssetDefinition instead of asset decorator

uranusjr committed:
Simplify group validators

The validate_identifier validator already checks the length, so we don't
need an extra one doing that.

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,155 @@
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        if "self" in parameters:
+            self.function: types.FunctionType = _handle_self_argument(cast(types.FunctionType, self.function))  # type: ignore[assignment]
```
Why do we need to do `cast(..., self.function)`? It is declared as `types.FunctionType`…

uranusjr replied:
And why does `function` need to be `FunctionType` at all, not just `Callable`?

Lee-W replied:
> Why do we need to do cast(..., self.function)? It is declared as types.FunctionType…

It's be classified as ClassDef by mypy. I "guess" it's due to the `self`

> And why does function need to be FunctionType at all, not just Callable?

It used to be `FunctionType`. I thought that was by design. 🤔 I can change it to `Callable`.

uranusjr replied:
It’s sort of by design because this should only be able to decorate a real function, not any callable (say lambda). But it’s not really that big a deal if the typing causes too many troubles elsewhere.

Lee-W replied:
Sounds good. I already make it `Callable` and this part seems to work well

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,155 @@
+        # TODO: Currently this just gets serialized into a string.
+        # When we create UI for assets, we should add logic to serde so the
+        # serialized DAG contains appropriate asset information.
+        dag._wrapped_definition = self
```
Is this still accurate? I see there’s a schema for `asset_definition` below.

(By the way, I am not even sure why I added this in the first place… this is not used anywhere. Maybe we can remove this altogether?)

Lee-W replied:
This is not used as of now. I thought this was designed for future use. But I would say it's better to remove it for now. I'll do it

Lee-W replied:
Removed

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,155 @@
+    group: str = ""
+    extra: dict[str, Any] = attrs.field(factory=dict)
+
+    def __call__(self, f: Callable) -> AssetDefinition:
```
Or, should this be `FunctionType` to match `AssetDefinition`?

Lee-W replied:
I kinda forget the deailts 🤔 But I think mypy complained about it. Let' me try again

Lee-W replied:
Just update all to `Callable`

uranusjr commented on airflow/models/asset.py:
```diff
@@ -37,8 +39,30 @@
+    from sqlalchemy.orm import Session
+
+
+@provide_session
```
We should not use `provide_session` here. The caller should always provide a session instead.

Lee-W replied:
Sure. Will remove it.

Lee-W replied:
removed

uranusjr commented on airflow/utils/context.py:
```diff
@@ -274,14 +279,27 @@ def __getitem__(self, key: int | str | Asset | AssetAlias) -> LazyAssetEventSele
                 raise IndexError(key)
+
+            if isinstance(obj, AssetRef):
+                obj = _fetch_active_assets_by_name([obj.name])
```
I wonder if this is a performance issue… when is `__getitem__` called?

Lee-W replied:
Yes, it might. I was thinking of whether we should fetch all the AssetRef when initializing. But that's kinda contradict the idea of lazy loading. Or maybe we could just load all the AssetRef when the first is loaded?

uranusjr replied:
Yeah maybe it’s easier to just load all AssetRef all at once.

Lee-W replied:
just updated to load it into `_assets` all at once

uranusjr commented on tests/always/test_example_dags.py:
```diff
@@ -199,6 +199,10 @@ def test_should_be_importable(example: str):
 def test_should_not_do_database_queries(example: str):
+    if example.endswith("airflow/example_dags/example_asset_decorator.py"):
+        # The asset decorator tries to write DAG and Asset into db which is expected
+        return
```
Not sure if we should do this… other examples don’t need to write the DAG and Asset into db?

Lee-W replied:
I thought it was the case 🤔 need to dig a bit to confirm.

uranusjr commented on tests/api_fastapi/core_api/routes/ui/test_assets.py:
```diff
@@ -47,5 +47,5 @@ def test_next_run_assets(test_client, dag_maker):
     assert response.json() == {
         "asset_expression": {"all": ["s3://bucket/key/1"]},
-        "events": [{"id": 17, "uri": "s3://bucket/key/1", "lastUpdate": None}],
+        "events": [{"id": 20, "uri": "s3://bucket/key/1", "lastUpdate": None}],
```
Why does this change?

Lee-W replied:
By the time the fixture `test_client` is used, the example DAGs are loaded.

https://github.com/apache/airflow/blob/28d0a7e90a5a56bc20ba582a0af355f53504aa28/tests/api_fastapi/core_api/routes/ui/test_assets.py#L38

and it loads the example DAG here

https://github.com/apache/airflow/blob/28d0a7e90a5a56bc20ba582a0af355f53504aa28/airflow/www/extensions/init_dagbag.py#L29

because `include_examples` is not set and then falls back to `conf.getboolean("core", "LOAD_EXAMPLES")`

Lee-W committed:
style(dag): remove _wrapped_definition

Lee-W committed:
style(decorators/assets): change types.FunctionType to Callable

Lee-W committed:
refactor(decorators/assets): make session in _fetch_active_assets_by_name required

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,151 @@
+                if not value:
+                    value = Asset(name=self._definition_name)
+                    if self._uri is not None:
+                        value.uri = self._uri
```
We probably should do better here. Theoratically, since this is only run at task execution time, the current asset should always be available in `active_assets` (which is from db populated at parse time). On the other hand, it kind of feels wrong to just create a dangling asset. Maybe we should just always assume the asset exists here.

Lee-W replied:
Yep, I think you're right. Just updated it

uranusjr commented on airflow/decorators/assets.py:
```diff
@@ -0,0 +1,151 @@
+                uri=self.uri,
+            )
+
+        DAG.bulk_write_to_db([dag])
```
Should we call this here? We should probably delegate to DagBag to do this writing instead. That’s where “normal” DAGs are written.

uranusjr replied:
Maybe this is the cause to the this? https://github.com/apache/airflow/pull/41325#discussion_r1832445526

Lee-W replied:
Yep, this is what I suspect as well 🤔 Will try it after the task_sdk pr

Lee-W replied:
removed and seems to work well

uranusjr commented on airflow/utils/context.py:
```diff
@@ -274,14 +279,27 @@ def __getitem__(self, key: int | str | Asset | AssetAlias) -> LazyAssetEventSele
                 raise IndexError(key)
+
+            if isinstance(obj, AssetRef):
+                obj = _fetch_active_assets_by_name([obj.name], self._session)
```
Is this correct? An IndexError is raised when `obj` is _not_ Asset or AssetAlias; how can it reach the next if below?

Lee-W replied:
Looks like it's wrong. I'm not sure why it seemed to work the last time I tried 🤔, but this is going to be redesigned for loading all the assets all at once. will take extra care on this issue

Lee-W replied:
just rewrote it

Lee-W committed:
fix(decorators/asets): remove DAG.bulk_write_to_db and remove self handling

Lee-W committed:
feat(utils/context): fetch asset_refs all at once

uranusjr commented on airflow/utils/context.py:
```diff
@@ -281,10 +289,13 @@ def __getitem__(self, key: int | str | Asset | AssetAlias) -> LazyAssetEventSele
             asset = self._assets[extract_event_key(obj)]
             join_clause = AssetEvent.asset
-            where_clause = AssetModel.uri == asset.uri
+            where_clause = AssetModel.name == asset.name
```
Hmm… I wonder if we should try to support _both_ name and URI here. Also this current code would not work because `extract_event_key` returns the URI.

Lee-W replied:
Right. This one is a bit tricky and intended to keep it this way for now. (This supports assets like `Asset(uri="...")` as the name fallbacks to its uri)

Lee-W replied:
I'm actually thinking of removing `str` support entirely. As we support more and more types, passing a string could be pretty confusing.

uranusjr replied:
That’s true. We’ll need to deprecate it in 2 though.

Lee-W replied:
If we agree to remove it, here's what I'm thinking.

1. Create another PR on `v2-10-test` that adds a deprecated warning for accessing the outlet through `str` 
2. Review how Assets are handled in outlets and create another PR or combine it with https://github.com/apache/airflow/pull/43774 (name, group, URI are not correctly handled in some places)

uranusjr replied:
Sounds good to me

Lee-W replied:
PR for 1 created https://github.com/apache/airflow/pull/43922
